clutch/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+                    cont_out: output.cont.fmt_to_string(store),
+                    status: output.status(),
+                    iterations: None,
+                });
+
+                self.last_claim = Some(claim);
+
+                Ok((input, output, iterations))
+            }
+            e => e,
+        }
+    }
+}
+
+impl<F: LurkField> ClutchState<F> {
+    fn hist(&self, n: usize) -> Option<&IO<F>> {
+        self.history.get(n)
+    }
+}
+
+impl ClutchState<F> {
+    fn commit(&mut self, store: &mut Store<F>, rest: Ptr<F>) -> Result<Option<Ptr<F>>> {
+        let (first, rest) = store.car_cdr(&rest)?;
+        let (second, _) = store.car_cdr(&rest)?;
+
+        let (expr, secret) = if rest.is_nil() {
+            // TODO: also support Commitment::from_ptr_with_hiding (randomized secret at runtime).
+            (first, F::zero())
+        } else if let Expression::Num(n) = store.fetch(&second).unwrap() {
+            (first, n.into_scalar())
+        } else {
+            bail!("Secret must be a Num")
+        };
+
+        let (evaled, _, _, _) = self.repl_state.eval_expr(expr, store)?;
+
+        let commitment = Commitment::from_ptr_and_secret(store, &evaled, secret);
+
+        let committed_expression = CommittedExpression {
+            expr: LurkPtr::from_ptr(store, &evaled),
+            secret: Some(secret),
+            commitment: Some(commitment),
+        };
+
+        self.expression_map.set(commitment, &committed_expression)?;
+        Ok(Some(store.intern_maybe_opaque_comm(commitment.comm)))
+    }
+
+    fn open(&mut self, store: &mut Store<F>, rest: Ptr<F>) -> Result<Option<Ptr<F>>> {
+        Ok(self.open_aux(store, rest)?.1)
+    }
+
+    fn apply_comm(
+        &mut self,
+        store: &mut Store<F>,
+        comm: Ptr<F>,
+        rest: Ptr<F>,
+    ) -> Result<Option<Ptr<F>>> {
+        let call_form = store.cons(comm, rest);
+        self.apply_comm_aux(store, call_form, false)
+    }
+    fn call(&mut self, store: &mut Store<F>, rest: Ptr<F>) -> Result<Option<Ptr<F>>> {
+        self.apply_comm_aux(store, rest, false)
+    }
+    fn chain(&mut self, store: &mut Store<F>, rest: Ptr<F>) -> Result<Option<Ptr<F>>> {
+        let x = self.apply_comm_aux(store, rest, true)?;
+
+        if let Some(cons) = x {
+            let (result_expr, new_comm) = store.car_cdr(&cons)?;
+
+            let new_secret0 = store.secret(new_comm).expect("secret missing");
+            let new_secret = *store
+                .get_expr_hash(&new_secret0)
+                .expect("hash missing")
+                .value();
+
+            let (_, new_fun) = store.open(new_comm).expect("opening missing");
+            let new_commitment = Commitment::from_comm(store, &new_comm);
+
+            let expr = LurkPtr::from_ptr(store, &new_fun);
+
+            let new_function = CommittedExpression::<F> {
+                expr,
+                secret: Some(new_secret),
+                commitment: Some(new_commitment),
+            };
+
```

Comment:
I think we can avoid the expects with something like this:
```rust
        let (result_expr, new_comm) = self
            .apply_comm_aux(store, rest, true)?
            .and_then(|cons| store.car_cdr(&cons).ok()) // unpack pair
            .ok_or_else(|| {
                anyhow!(
                    "chained functional commitment output must be a pair (result, new-commitment)"
                )
            })?;

        let new_secret = store
            .secret(new_comm)
            .ok_or_else(|| anyhow!("secret missing"))
            .and_then(|hash| {
                store
                    .get_expr_hash(&hash)
                    .ok_or_else(|| anyhow!("hash missing"))
            })
            .and_then(|hash| Ok(hash.value().clone()))?;

        let (_, new_fun) = store
            .open(new_comm)
            .ok_or_else(|| anyhow!("opening missing"))?;
        let new_commitment = Commitment::from_comm(store, &new_comm);
        let expr = LurkPtr::from_ptr(store, &new_fun);

        let new_function = CommittedExpression::<F> {
            expr,
            secret: Some(new_secret),
            commitment: Some(new_commitment),
        };

```

Note: incomplete snippet, the rest is mostly the same but the final bail goes away.